src/typescript/sdk/src/markets/utils.ts

```diff
@@ -322,3 +328,111 @@ export function calculateTvlGrowth(periodicStateTracker1D: Types["PeriodicStateT
   // (b * c) / (a * d)
   return b.mul(c).div(a.mul(d)).toString();
 }
+
+type ReservesAndBondingCurveState = Flatten<
+  AtLeastOne<{
+    inBondingCurve: boolean;
+    lpCoinSupply: bigint;
+  }> & {
+    clammVirtualReserves: Types["Reserves"];
+    cpammRealReserves: Types["Reserves"];
+  }
+>;
+
+/**
+ * Calculates the circulating supply based on the given market state.
+ *
+ * The logic for calculation is taken directly from the Move smart contract.
+ * @see assign_supply_minuend_reserves_ref_mut in `emojicoin_dot_fun.move`
+ */
+export const calculateCirculatingSupply = ({
+  clammVirtualReserves,
+  cpammRealReserves,
+  ...args
+}: ReservesAndBondingCurveState) =>
+  isInBondingCurve(args)
+    ? BASE_VIRTUAL_CEILING - clammVirtualReserves.base
+    : EMOJICOIN_SUPPLY - cpammRealReserves.base;
```

Comment:
- [ ] Construct test where a single user makes market buys in bonding curve, sells in bonding curve, to exact transition, and past bonding curve, then sells